### PR TITLE
Fix missing touch

### PR DESCRIPTION
**Pull request by harendra-kumar**
Adding a touchForeignPtr led to significant regressions. Fixing that required a bigger change to use a simpler type in the array instead of using ForeignPtr.

### CODE REVIEW

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 1]
```diff
@@ -300,12 +308,39 @@ memcmp p1 p2 len = do
     r <- c_memcmp p1 p2 (fromIntegral len)
     return $ r == 0
 
--- | Same as unsafeWithForeignPtr but lifted to any MonadIO monad.
-{-# INLINE unsafeWithForeignPtrM #-}
-unsafeWithForeignPtrM :: MonadIO m => ForeignPtr a -> (Ptr a -> m b) -> m b
-unsafeWithForeignPtrM fo f = do
-  r <- f (unsafeForeignPtrToPtr fo)
-  liftIO $ touchForeignPtr fo
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:300]: Why remove this?

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 2]
```diff
@@ -244,20 +259,13 @@ import qualified Streamly.Internal.Data.Fold.Type as FL
...
 import Prelude hiding
     (length, foldr, read, unlines, splitAt, reverse, truncate)
 
-#if MIN_VERSION_base(4,10,0)
-import Foreign.ForeignPtr (plusForeignPtr)
-#else
-import GHC.Base (Int(..), plusAddr#)
-import GHC.ForeignPtr (ForeignPtr(..))
-plusForeignPtr :: ForeignPtr a -> Int -> ForeignPtr b
-plusForeignPtr (ForeignPtr addr c) (I# d) = ForeignPtr (plusAddr# addr d) c
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:250]: Is there something incorrect here?

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 3]
```diff
@@ -300,12 +308,39 @@ memcmp p1 p2 len = do
...
+fptrToArrayContents :: ForeignPtrContents -> ArrayContents
+arrayToFptrContents :: ArrayContents -> ForeignPtrContents
+#ifdef USE_FOREIGN_PTR
+fptrToArrayContents = ArrayContents
+arrayToFptrContents (ArrayContents contents) = contents
+#else
+fptrToArrayContents (PlainPtr mbarr) = ArrayContents mbarr
+fptrToArrayContents _ = error "Unsupported foreign ptr"
+arrayToFptrContents (ArrayContents contents) = PlainPtr contents
+#endif
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:335]: Can you please elaborate on this?

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 4]
```diff
@@ -634,6 +699,11 @@ arrayChunkBytes = 1024
 -- Snoc
 -------------------------------------------------------------------------------
 
+{-# INLINE touch #-}
+touch :: ArrayContents -> IO ()
+touch (ArrayContents contents) =
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:704]: Add this on the top of the file? Define before use?

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 5]
```diff
@@ -343,8 +378,8 @@ data Array a =
     Storable a =>
 #endif
     Array
-    { arrContents :: {-# UNPACK #-} !ForeignPtrContents -- ^ first address
-    , arrStart :: {-# UNPACK #-} !(Ptr a)
+    { arrContents :: UNPACKIF !ArrayContents
+    , arrStart :: {-# UNPACK #-} !(Ptr a)      -- ^ first address
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs]: Why this change?

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 6]
```diff
@@ -1155,24 +1177,24 @@ arraysOf n (D.Stream step state) =
...
-            D.Stop -> return $ D.Skip (GroupYield start end bound GroupFinish)
+                    then D.Skip (GroupYield contents start end' bound (GroupStart s))
+                    else D.Skip (GroupBuffer s contents start end' bound)
+            D.Skip s -> return $ D.Skip (GroupBuffer s contents start end bound)
+            D.Stop -> return $ D.Skip (GroupYield contents start end bound GroupFinish)
 
-    step' _ (GroupYield start end bound next) =
-        return $ D.Yield (Array start end bound) next
+    step' _ (GroupYield contents start end bound next) =
+        return $ D.Yield (Array contents start end bound) next
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1271]: Don't we ever touch foreign ptr ever here? (Question for the original code)

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 7]
```diff
@@ -1355,7 +1391,7 @@ toListFB c n Array{..} = go (unsafeForeignPtrToPtr aStart)
         -- XXX
         let !x = unsafeInlineIO $ do
                     r <- peek p
-                    touchForeignPtr aStart
+                    touch arrContents
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1452]: `touchFPContents` vs `touch`.

File: src/Streamly/Internal/Data/Array/Foreign/Type.hs [hunk 8]
```diff
@@ -223,7 +226,8 @@ fromPtr n ptr = unsafeInlineIO $ do
     fptr <- newForeignPtr_ ptr
     let end = ptr `plusPtr` n
     return $ Array
-        { aStart = fptr
+        { arrStart = ptr
+        , arrContents = undefined
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Foreign/Type.hs]: What happens when you `touch` `undefined`?
If we use `fromPtr`, will we be able to use the result in other combinators?

File: src/Streamly/Internal/Unicode/Stream.hs [hunk 9]
```diff
@@ -437,8 +437,7 @@ replacementChar = '\xFFFD'
 decodeUtf8WithD :: Monad m
     => CodingFailureMode -> Stream m Word8 -> Stream m Char
 decodeUtf8WithD cfm (Stream step state) =
-    let A.Array p _ = utf8d
-        !ptr = unsafeForeignPtrToPtr p
+    let A.Array contents ptr _ = utf8d
```
Comment (adithyaov) [inline src/Streamly/Internal/Unicode/Stream.hs]: Perform `!ptr = p` like you did above? Is that necessary?

**Review by adithyaov — APPROVED** [review]
I did not see any obvious error. LGTM

File: src/Streamly/Internal/Data/Array/Foreign/Type.hs [hunk 10]
```diff
@@ -139,7 +138,11 @@ data Array a =
     Storable a =>
 #endif
     Array
-    { arrContents :: {-# UNPACK #-} !ForeignPtrContents -- first address
+    { arrContents ::
+#ifndef USE_FOREIGN_PTR
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Foreign/Type.hs:142]: Don't you need to initialize the flag accordingly here?

Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Array/Foreign/Type.hs:142]: Remains undefined by default.

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs (hunk 2 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:250]: Looks ok to me.

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs (hunk 4 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:704]: Moved.

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs (hunk 6 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1271]: Looks like a bug. But may not occur in practice depending on how the array is used. We can try adding a touch and check if it degrades perf.
